Blogpost about Python runtime (#600)
home-assistant/developers.home-assistant: blog/2020-07-13-alpine-python.md

---
author: Pascal Vizeli
authorURL: https://github.com/pvizeli
authorTwitter: pvizeli
title: "Alpine with musl and Python"
---

> It is well known that we use Alpine for most of our Containers. It is the perfect distribution for containers because it is small (BusyBox based), available for a lot of CPU architectures, and the package system is slim. Alpine and musl are new and had at least a significant development progress. That means many statements are today just wrong. I will clean up this incorrect information based on my research to provide the fastest way to run Home Assistant. This blogpost is not a musl vs. glibc flame ware; we use glibc on our OS. I will clean up old or half-truth, which is too long now in people's minds.

Reviewer: Is it well known or should this be

We use Alpine for most of our containers.

Reviewer: > Alpine and musl are new and had at least a significant development progress. That means many statements are today just wrong. I will clean up this incorrect information based on my research to provide the fastest way to run Home Assistant.

Alpine and musl are relatively new and have made a lot of progress since their starts. Because of this, many benchmarks out there are show their old performance, while they perform a lot better today.

With the latest optimizations, this combination is really fast. To show how fast, I've run some benchmarks against our Python images.

Author: Musl was released 9 years ago which is young for an c library in my interpretation. I don't know the full history of Alpine. It's more in container aspect.

> For tests, I used the images from [Docker Python library](https://github.com/docker-library/python), and the result is published to our [base images](https://github.com/home-assistant/docker-base). I used pyperformance for labs testing and the Home Assistant internal benchmark tools for more Real-life information. The test environment is running inside a container on the same docker host.

Reviewer: real-life

## C/POSIX standard library

> I read often: Python is slower if they use musl as default C library. This fact is not 100% correct. If both Python runtime was compiled with the same GCC and with `-O3`, the glibc variant is a bit faster in the lab benchmark, but in the real world, it makes not a difference. Alpine compile it for `-Os` and most distro with `-O2`, which gives the often written difference between the Python runtime interpreters, but with the same compiler optimizing, musl based Python runtimes have no negative side effect.

Reviewer: when using musl

> But there is a game-changer which makes the musl one more useful as the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, jemalloc is compatible today with musl, and there was a time which it was not.

Reviewer 2: ```suggestion
But there is a game-changer, which makes the musl one more useful compared to the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect, which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, today's jemalloc is compatible with musl (there was a time which it was not).
```

Reviewer 3: > There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without.

IMO this could benefit from re-phrasing. Frankly, I've read it quite a few times but still don't quite understand what "setups" are being compared here. And it could be just me...

for example `that musl is much faster as glibc` if it is faster as in "comparison" then we should probably use "than" as in `musl is much faster than glibc` or if you meant they offer similar speeds then it should be `that musl is as fast as glibc`

> but glibc is slower with jemalloc as without

Same here. is it just me having issues parsing it? (wouldn't be the 1st time :)

Reviewer 3: > but glibc is slower with jemalloc as without

but glibc with jemalloc is slower than glibc without it.

Reviewer 3: so as I understand, we are comparing:
1. musl with jemalloc
2. glibc with jemalloc
3. glibc without jemalloc
so we just should rank the above?

## Compiler

> How do you compile Python is also essential. There were at least statements from Fedora or Redhat about disable `semantic-interposition` to become a high-performance boost. I was not able to reproduce this on GCC 9.3.0, but I also saw no adverse side effects. I can recommend disabling the semantics like the built-in allocator optimization and link jemalloc at build time. I will also recommend using the `-O3` optimization. We never saw an issue with this aggressive optimization on our target platforms. I need to say, unlike the distro Python runtime interpreters, we don't need to run everywhere. So we can use the `--enable-optimizations` without any overwrite and add more flags. I can say today, PGO/LTO/O3 make Python faster and at least work on our target CPUs.

Reviewer 2: ```suggestion
How you compile Python is also essential. There were statements from Fedora or Redhat about disable `semantic-interposition` to get a high-performance boost. I was not able to reproduce this on GCC 9.3.0, but I also saw no adverse side-effects. I can recommend disabling the semantics like the built-in allocator optimization and link jemalloc at build time. I will also recommend using the `-O3` optimization. We never saw an issue with this aggressive optimization on our targeted platforms. I need to say, unlike the distro Python runtime interpreters, we don't need to run everywhere. So we can use the `--enable-optimizations` without any overwrite and add more flags. I can say today, PGO/LTO/O3 make Python faster and it works on our target CPUs.
```

## Python packages

> Yes, Alpine indeed has no manylinux compatibility with musl. If you don't cache your builds, it needs every time compiling the C extensions on that packages. This process would take time. More time if you also crossbuild with Qemu for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a loose there because the packages are mostly also not optimized for the target system. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and hold it up to date over CI agents. For sure, not every project can take this effort, but also not all repository haver over 1k packages for each CPU architecture, and the build time of the docker file is not so important at all.

Reviewer 2: ```suggestion
Yes, Alpine indeed has no manylinux compatibility with musl. If you don't cache your builds, it needs to compile the C extensions on those packages. This process would take time, even time if you also cross-build with Qemu for different CPU architectures. You cannot get precompiled binaries from PyPi. I don't see a loss there because the packages are, mostly, also not optimized for the target system. We created our wheel [index](https://wheels.home-assistant.io/) and [backend](https://github.com/home-assistant/wheels) to compile and hold it up to date using CI agents. For sure, not every project can take this effort, but also no repository has pre-build over 1k packages for each CPU architecture, and the build time of the Docker file is not so important at all.
```

> Yes, Alpine indeed has no manylinux compatibility with musl. If you don't cache your builds, it needs every time compiling the C extensions on that packages. This process would take time. More time if you also crossbuild with Qemu for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a loose there because the packages are mostly also not optimized for the target system. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and hold it up to date over CI agents. For sure, not every project can take this effort, but also not all repository haver over 1k packages for each CPU architecture, and the build time of the docker file is not so important at all.

Reviewer 4: ```suggestion
Alpine Linux has no [manylinux](https://github.com/pypa/manylinux) compatibility with musl. If you don't cache your builds then it's required to compiling the C extensions for that packages every time. This process takes time. More time if you also crossbuild with QEMU for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a downside here because the packages are mostly not optimized for the target systems. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and keep it up to date with CI agents. For sure, not every project can take this effort, but also not all repository have over 1000 packages for each CPU architecture. The build time of the Docker file is not the limiting factor at all.
```

## Alpine Linux

> It is a great base system for Container and allows us to provide the best experience to our User. A big thanks to Alpine Linux, musl, and jemalloc, which make that possible.
> 
> Here is a Benchmark between the distro Python and our optimization (GCC 9.3.0 / musl / Python 3.8.3)

Reviewer 2: ```suggestion
It is a great base system for Container and allows us to provide the best experience to our user. A big thanks to Alpine Linux, musl, and jemalloc, which make this all possible.

Here is a Benchmark between the distro Python and our optimization (GCC 9.3.0 / musl / Python 3.8.3)
```

> Here is a Benchmark between the distro Python and our optimization (GCC 9.3.0 / musl / Python 3.8.3)

Reviewer 4: ```suggestion
The table shows the results between the Alpine Linux's Python runtime and our optimization (GCC 9.3.0/musl/Python 3.8.3).
```

IMHO, there should be a little more context here (how the results were gathered, how it can be reproduced, how the selection was made, etc.). The naked table is hard to interpret, especially as at least one run is slower.

Author: We can also remove it 🤷‍♂️
